proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,725 @@
+individually. But for now, one immediate benefit of instance imports is that
+they allow for module and export names to be maximally factored out for both
+text- and binary-format reuse.
+
+
+### Module Imports and Nested Instances
+
+Symmetric to instances, modules can be imported via module type. Once a module
+is imported, it must be instantiated before it can be executed. This is achieved
+by creating *nested instances* via `instance` definitions. For example, in this
+code:
+```wasm
+(module
+  (import "m" (module $M
+    (import "in" (func))
+    (export "out" (func $out))
+  ))
+  (import "f" (func $f))
+  (instance $i (instance.instantiate $M (ref.func $f)))
+  (func (export "run")
+    call $i.$out
+  )
+)
+```
+the outer module imports a module `$M` and a function `$f` and then uses `$f` to
+instantiate `$m` producing an instance `$i`. `instance` definitions are created
+using [constant initializer expression], similar to `global` definitions. The
+`instance.instantiate` instruction introduced by this proposal is *only* allowed
+in constant initializer expressions, but could in the future be allowed as a
+normal instruction to enable runtime instantiation. Its signature is:
+```
+instance.instantiate $module : [ (ref Tᵢ)ⁿ ] -> [ (ref $InstanceT) ]
+```
+where
+* `$module` refers to a module (in the module index space) of type `$ModuleT`
+* the `n` operands `Tᵢ` match the `n` imports of `$ModuleT` in import vector order
+* `$InstanceT` is the instance type of `$ModuleT`
+
+Note that while `instance.instantiate` relies on the positional order of imports
+in the local type definition of `$ModuleT`, module subtyping is independent of
+order (exports are matched by string) and thus reordering the exports of a
+module definition will never break clients.
+
+Instance imports and definitions can also be used as operands to
+`instance.instantiate`. For example, this module imports a `wasi_file`
+instance and passes it on to its child:
+```wasm
+(module
+  (type $WasiFile (instance
+    (export "read" (func (param i32 i32 i32) (result i32)))
+    (export "write" (func (param i32 i32 i32) (result i32)))
+  ))
+  (import "wasi_file" (instance $wasi-file (type $WasiFile)))
+  (import "child" (module $child
+    (import "wasi_file" (instance (type $WasiFile)))
+  ))
+  (instance (instance.instantiate $child (ref.instance $wasi-file)))
+)
+```
+Here, `ref.instance` is a new instruction symmetric to [`ref.func`], selecting an
+instance from the instance index space via a static index immediate. It will be
+discussed more in the next section.
+
+`instance.instantiate` allows its operands to be subtypes of the declared
+module import types. When the operands are of instance or module type, that
+means checking module/instance subtyping. Unlike with [GC types], where
+subtyping needs to imply memory layout, module and instance subtyping can be
+order-independent, matching up by import/export names instead.
```

Comment:
Q: is this possible:

```wasm
(module
  (type $WasiFile (instance
    (export "read" (func (param i32 i32 i32) (result i32)))
    (export "write" (func (param i32 i32 i32) (result i32)))
  ))
  (import "wasi_file" (instance $wasi-file (type $WasiFile)))
  (import "child" (module $child
    (import "wasi_file" "read" (func (param i32 i32 i32) (result i32)))
    (import "wasi_file" "write" (func (param i32 i32 i32) (result i32)))
  ))
  (instance (instance.instantiate $child (ref.instance $wasi-file)))
)
```

Reply:
Yep, two-level imports are basically equivalent to a one-level import of a corresponding instance type.

Reply:
(Actually this is now being discussing with Rossberg [below](https://github.com/WebAssembly/module-types/pull/3#discussion_r419998951).)

Reply:
We resolved "yes": the two-level form is desugared to the one-level form.